File: channels/src/main/kotlin/com/tilal6991/channels/base/ChannelsApplication.java
package com.tilal6991.channels.base;

import android.app.Application;
import android.databinding.BindingMethod;
import android.databinding.BindingMethods;
import android.util.Log;
import android.view.View;

import com.squareup.leakcanary.LeakCanary;
import com.tilal6991.channels.inject.ChannelsObjectProvider;
import com.tilal6991.channels.inject.DaggerChannelsObjectProvider;
import com.tilal6991.channels.inject.RelayModule;
import timber.log.Timber;

@BindingMethods({
        @BindingMethod(type = View.class,
                attribute = "android:onClick",
                method = "setOnClickListener")
})
public class ChannelsApplication extends Application {

    ChannelsObjectProvider provider;

    public void onCreate() {
        super.onCreate();

        LeakCanary.install(this);

        provider = DaggerChannelsObjectProvider.builder()
                .relayModule(new RelayModule(this))
                .build();
        Timber.plant(new Timber.DebugTree() {
            @Override
            protected void log(int priority, String tag, String message, Throwable t) {
                if (priority == Log.ERROR) {
                    if (t != null) {
                        throw new RuntimeException(t);
                    } else {
                        throw new IllegalStateException(message);
                    }
                }
                super.log(priority, tag, message, t);
            }
        });
    }
}

File: channels/src/main/kotlin/com/tilal6991/channels/base/ChannelsApplication.kt
package com.tilal6991.channels.base

import android.app.Application
import com.brianegan.bansa.BaseStore
import com.tilal6991.channels.db.ConnectionDatabase
import com.tilal6991.channels.redux.*
import com.tilal6991.channels.viewmodel.RelayVM
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import timber.log.Timber

class ChannelsApplication : Application() {

    private val relayMiddleware = relayMiddleware(this)
    val store = RxStore(BaseStore(initialState, reducer, relayMiddleware), Schedulers.immediate())
    val relayHandle: RelayVM by lazy { RelayVM(this) }

    val state = store.observable()
            // .onBackpressureDrop()
            // .observeOn(AndroidSchedulers.mainThread())
            .share()

    override fun onCreate() {
        super.onCreate()

        val dbSubs = ConnectionDatabase.instance(this)
                .getConfigurations()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { store.dispatch(Actions.NewConfigurations(it)) }

        // LeakCanary.install(this)
        Timber.plant(Timber.DebugTree())
    }
}
